### PR TITLE
LTO fixes

### PR DESCRIPTION
**Pull request by kaspar030**
Based on #3361.

This PR adds the possibility to exclude some files from being built with LTO, and applies this to some critical files on Cortex-M.

Some boards don't compile (all msp430, mulle, pba-d-01-kw2x), but as this is completely optional (have to manually add LTO=1 to the Makefile, still get a big fat warning), and the codesize savings are 5-10%, IMHO we should get this merged as is.

### CODE REVIEW

**Comment by kaspar030** [comment]
@gebart Are you fine with me basically adopting your PR? I dropped the "LTO=yes" compatibility stuff and the WIP cortex fix commit from 3361.

**Comment by gebart** [comment]
@kaspar030 sure, feel free to improve on this. I agree with dropping the
LTO=yes argument btw.

Den 10 aug. 2016 4:40 PM skrev "Kaspar Schleiser" <notifications@github.com

File: cpu/cortexm_common/thread_arch.c [hunk 1]
```diff
@@ -332,11 +327,14 @@ __attribute__((naked)) void arch_context_switch(void)
     "ldr    r1, =sched_active_thread  \n" /* load address of current tcb */
     "ldr    r1, [r1]                  \n" /* dereference pdc */
     "str    r0, [r1]                  \n" /* write r0 to pdc->sp */
+    );
+}
+
+void __attribute__((naked)) __attribute__((used)) isr_svc(void) {
```
Comment (gebart) [inline cpu/cortexm_common/thread_arch.c]: Is it not dangerous to assume that these two functions will be placed in order and next to each other without any padding in the final binary? I'm sensing that we are relying on linker implementation specific behaviour here

Comment (kaspar030) [inline cpu/cortexm_common/thread_arch.c]: totally, I forgot the function call in pendsv. will fix.

Comment (kaspar030) [inline cpu/cortexm_common/thread_arch.c]: fixed

File: Makefile.vars [hunk 2]
```diff
@@ -35,6 +35,7 @@ export AS                    # The assembler.
 export ASFLAGS               # Flags for the assembler.
 export LINK                  # The command used to link the files. Must take the same parameters as GCC, i.e. "ld" won't work.
 export LINKFLAGS             # Flags to supply in the linking step.
+export LTOFLAGS              # Pass link time optimization flags to submake
```
Comment (gebart) [inline Makefile.vars]: I don't understand what this does from reading the comment in the Makefile. Is this a variant of CFLAGS?

Comment (kaspar030) [inline Makefile.vars]: >> +export LTOFLAGS              # Pass link time optimization flags to submake
> 
> I don't understand what this does from reading the comment in the Makefile. Is this a variant of CFLAGS?

What about "extra CFLAGS for compiling with link time optimization?

Comment (gebart) [inline Makefile.vars]: much better

**Comment by kaspar030** [comment]
- improved LINKFLAGS comment text

File: cpu/cortexm_common/thread_arch.c [hunk 3]
```diff
@@ -332,11 +327,15 @@ __attribute__((naked)) void arch_context_switch(void)
...
     "ldr    r1, [r1]                  \n" /* dereference pdc */
     "str    r0, [r1]                  \n" /* write r0 to pdc->sp */
+    "bl     isr_svc                   \n" /* continue with svc */
+    );
+}
+
+void __attribute__((naked)) __attribute__((used)) isr_svc(void) {
+    __asm__ volatile (
     /* SVC handler entry point */
     /* PendSV will continue from above and through this part as well */
```
Comment (gebart) [inline cpu/cortexm_common/thread_arch.c]: Reword or remove this comment.

File: cpu/ezr32wg/Makefile [hunk 4]
```diff
@@ -4,4 +4,7 @@ MODULE = cpu
 # add a list of subdirectories, that should also be build
 DIRS = periph $(RIOTCPU)/cortexm_common
 
+#
```
Comment (gebart) [inline cpu/ezr32wg/Makefile]: Forgot the comment text?

Comment (kaspar030) [inline cpu/ezr32wg/Makefile]: I use empty comment lines to logically seperate lines from the comment above. Do you think it is necessary to add "ignore file for link time optimization" is necessary for all these?

File: cpu/cortexm_common/thread_arch.c (hunk 3 above)
Comment (kaspar030) [inline cpu/cortexm_common/thread_arch.c]: reworded

File: cpu/ezr32wg/Makefile (hunk 4 above)
Comment (gebart) [inline cpu/ezr32wg/Makefile]: it looks like something is missing without a text.
What happens when you compile vectors.c with LTO? Do you know if it is enough to add a attribute used tags somewhere or if there's more to it?

Comment (kaspar030) [inline cpu/ezr32wg/Makefile]: > What happens when you compile vectors.c with LTO?
> Do you know if it is enough to add a attribute used tags somewhere
> or if there's more to it?

Somehow it triggers this error:

    lto1: internal compiler error: in add_symbol_to_partition_1, at
lto/lto-partition.c:135
    Please submit a full bug report,
    with preprocessed source if appropriate.
    See <https://bugs.archlinux.org/> for instructions.
    lto-wrapper: fatal error: arm-none-eabi-gcc returned 1 exit status
    compilation terminated.
    /usr/lib/gcc/arm-none-eabi/6.1.1/../../../../arm-none-eabi/bin/ld:
error: lto-wrapper failed
    collect2: error: ld returned 1 exit status
    make: ***
[/home/kaspar/src/riot.tmp/examples/default/../../Makefile.include:261:
all] Error 1

**Comment by kaspar030** [comment]
- added comments to ```SRC_NOLTO``` Makefile statements

File: cpu/ezr32wg/Makefile (hunk 4 above)
Comment (kaspar030) [inline cpu/ezr32wg/Makefile]: > it looks like something is missing without a text.

I've added a comment with link to corresponding github issue.

**Comment by kaspar030** [comment]
@gebart IMHO this can be merged, what do you think?

**Comment by gebart** [comment]
I agree. Since LTO is still experimental it does not matter that it only produces working binaries for some of the platforms.

**Comment by kaspar030** [comment]
- squashed

> I agree. Since LTO is still experimental it does not matter that it only produces working binaries for some of the platforms.

Is that an ACK? :)

**Comment by gebart** [comment]
Changes look fine, though it's missing SRC_NOLTO for kw2x and k60 (vector.c) if that causes the same build error as the other cortex vectors.

If you feel like it you could rename vector.c -> vectors.c in kw2x and k60 to match all the other Cortex CPUs.

**Comment by kaspar030** [comment]
> If you feel like it you could rename vector.c -> vectors.c in kw2x and k60 to match all the other Cortex CPUs.

did!

**Comment by gebart** [comment]
ACK

**Comment by gebart** [comment]
Murdock is happy -> go!

**Pull request merged**